### PR TITLE
feat(db): delete deployments

### PR DESCRIPTION
**Pull request by ig-ra**
ENG-123

### CODE REVIEW

File: backend/internal/db/dbgorm/deployments.go [hunk 1]
```diff
@@ -43,22 +43,24 @@ func (db *gormdb) getDeployment(ctx context.Context, deploymentID string) (*sche
...
-		return translateError(err)
-	}
-	// FIXME: delete deployment sessions as well?
+func (db *gormdb) deleteDeployment(ctx context.Context, deploymentID string) error {
+	// TODO: this won't work. Consider removing generic delete function.
+	// return delete(db.db, ctx, scheme.Deployment{}, "deployment_id = ?", deploymentID)
 
-	return nil
+	var b scheme.Deployment = scheme.Deployment{DeploymentID: deploymentID}
+	return db.db.WithContext(ctx).Delete(&b).Error
```
Comment (efiShtain) [inline backend/internal/db/dbgorm/deployments.go]: did we decide on cascading deletion?
deployments should delete all of its sessions (and session all of its sub types)?

File: backend/internal/db/dbgorm/deployments.go [hunk 2]
```diff
@@ -43,22 +43,24 @@ func (db *gormdb) getDeployment(ctx context.Context, deploymentID string) (*sche
...
 
-func (db *gormdb) DeleteDeployment(ctx context.Context, id sdktypes.DeploymentID) error {
-	var b scheme.Deployment
-	if err := db.db.WithContext(ctx).Where("deployment_id = ?", id.String()).Delete(&b).Error; err != nil {
-		return translateError(err)
-	}
-	// FIXME: delete deployment sessions as well?
+func (db *gormdb) deleteDeployment(ctx context.Context, deploymentID string) error {
+	// TODO: this won't work. Consider removing generic delete function.
+	// return delete(db.db, ctx, scheme.Deployment{}, "deployment_id = ?", deploymentID)
```
Comment (efiShtain) [inline backend/internal/db/dbgorm/deployments.go]: don't really this need comment

File: backend/internal/db/dbgorm/builds.go [hunk 3]
```diff
@@ -10,18 +10,18 @@ import (
 	"go.autokitteh.dev/autokitteh/sdk/sdktypes"
 )
 
+func (db *gormdb) saveBuild(ctx context.Context, build scheme.Build) error {
```
Comment (efiShtain) [inline backend/internal/db/dbgorm/builds.go]: I assume you just saw this one and fixed it which is great
but if this is unrelated to this PR, please leave a comment on the PR itself or split it to a different PR
since It takes me time to understand why you did it

File: backend/internal/db/dbgorm/deployments.go [hunk 4]
```diff
@@ -43,22 +43,24 @@ func (db *gormdb) getDeployment(ctx context.Context, deploymentID string) (*sche
...
+	var b scheme.Deployment = scheme.Deployment{DeploymentID: deploymentID}
+	return db.db.WithContext(ctx).Delete(&b).Error
 }
 
-// FIXME: fix generic in order to avoid all this
+func (db *gormdb) DeleteDeployment(ctx context.Context, deploymentID sdktypes.DeploymentID) error {
+	return translateError(db.deleteDeployment(ctx, deploymentID.String()))
+}
+
+// TODO: rewrite generic in order to avoid splitting to Deployment and DeploymentWithStats
```
Comment (efiShtain) [inline backend/internal/db/dbgorm/deployments.go]: TODO without a ticket would probably never be solved, 
so better to add a ticket and list it here like TODO(ENG-XYZ): ...

File: backend/internal/db/dbgorm/gorm_test.go [hunk 5]
```diff
@@ -41,30 +49,23 @@ func setupDB(dbName string) *gorm.DB {
...
-			log.Printf("Could not sync logger: %v", err)
-		}
-	}()
-	return logger
-}
-
-func newDbFixture() *dbFixture {
+func newDbFixture(withoutForeignKeys bool) *dbFixture {
 	db := setupDB("") // in-memory db, specify filename to use file db
+	if withoutForeignKeys {
```
Comment (efiShtain) [inline backend/internal/db/dbgorm/gorm_test.go]: when would you use this ?

File: backend/internal/db/dbgorm/gorm_test.go [hunk 6]
```diff
@@ -41,30 +49,23 @@ func setupDB(dbName string) *gorm.DB {
 		},
 		Logger: logger,
 	})
+	db.Exec("PRAGMA foreign_keys = ON")
```
Comment (efiShtain) [inline backend/internal/db/dbgorm/gorm_test.go]: you have to reimplement this logic here since you are not using our gorm wrapper
maybe you should consider using ours

File: backend/internal/db/dbgorm/gorm_test.go [hunk 7]
```diff
@@ -82,39 +83,61 @@ func newDbFixture() *dbFixture {
...
+func newBuild() scheme.Build {
+	now := time.Now().UTC() // save and compare times in UTC
+
+	return scheme.Build{
+		BuildID:   testBuildID,
+		Data:      []byte{},
+		CreatedAt: now,
+	}
+}
+*/
```
Comment (efiShtain) [inline backend/internal/db/dbgorm/gorm_test.go]: remove comments

File: backend/internal/db/dbgorm/scheme/scheme.go [hunk 8]
```diff
@@ -370,6 +370,10 @@ type Deployment struct {
 	State        int32
 	CreatedAt    time.Time
 	UpdatedAt    time.Time
+	DeletedAt    gorm.DeletedAt `gorm:"index"`
+
+	// just for foreign key constraint. Without it gorm won't enforce it
+	Sessions []*Session `gorm:"foreignKey:DeploymentID"`
```
Comment (efiShtain) [inline backend/internal/db/dbgorm/scheme/scheme.go]: could you clarify this one?

File: backend/internal/db/dbgorm/scheme/scheme.go [hunk 9]
```diff
@@ -322,7 +322,7 @@ func ParseSessionCallAttemptComplete(c SessionCallAttempt) (d sdktypes.SessionCa
 
 type Session struct {
 	SessionID        string `gorm:"primaryKey"`
-	DeploymentID     string `gorm:"index"`
+	DeploymentID     string `gorm:"index;foreignKey:DeploymentID;references:DeploymentID"`
```
Comment (efiShtain) [inline backend/internal/db/dbgorm/scheme/scheme.go]: do you need both foreignKey and references here?

File: backend/internal/db/dbgorm/scheme/scheme.go [hunk 10]
```diff
@@ -388,6 +392,11 @@ func ParseDeployment(d Deployment) (sdktypes.Deployment, error) {
 	return deployment, nil
 }
 
+// gorm don't cascade soft deletes. hook is transactional.
+func (d *Deployment) AfterDelete(db *gorm.DB) (err error) {
+	return db.Where("deployment_id = ?", d.DeploymentID).Delete(&Session{}).Error
```
Comment (efiShtain) [inline backend/internal/db/dbgorm/scheme/scheme.go]: what about running sessions ?
we don't allow deleting them
we should probably not allow deleting a deployment with a running session

other option is deleting but letting the running session to complete or fail

File: backend/internal/db/dbgorm/sessions.go [hunk 11]
```diff
@@ -17,16 +17,24 @@ import (
...
-func (db *gormdb) GetSession(ctx context.Context, sessionID sdktypes.SessionID) (sdktypes.Session, error) {
-	return getOneWTransform(db.db, ctx, scheme.ParseSession, "session_id = ?", sessionID.String())
+func (db *gormdb) getSession(ctx context.Context, sessionID string) (*scheme.Session, error) {
+	return getOne(db.db, ctx, scheme.Session{}, "session_id = ?", sessionID)
+}
+
+func (db *gormdb) GetSession(ctx context.Context, id sdktypes.SessionID) (sdktypes.Session, error) {
+	s, err := db.getSession(ctx, id.String())
+	if s == nil || err != nil {
+		return nil, translateError(err)
```
Comment (efiShtain) [inline backend/internal/db/dbgorm/sessions.go]: what does translate error of nil return ?

File: backend/internal/deployments/deployments.go [hunk 12]
```diff
@@ -150,6 +150,27 @@ func (d *deployments) Drain(ctx context.Context, id sdktypes.DeploymentID) error
...
+		}
+
+		if sdktypes.GetDeploymentState(dep) != sdktypes.DeploymentStateInactive {
+			err = drain(ctx, tx, id)
+		}
+		return err
+	})
+
+	// NOTE: deployment is inactive, so it should be safe to delete it outside of the transaction.
+	// It seems that sqlite does not support nested transactions, so we'll fail in delete hook, since DB is locked.
```
Comment (efiShtain) [inline backend/internal/deployments/deployments.go]: so... what is the solution here?

File: cmd/ak/cmd/deployments/delete.go [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+import (
+	"github.com/spf13/cobra"
+
+	"go.autokitteh.dev/autokitteh/cmd/ak/common"
+	"go.autokitteh.dev/autokitteh/internal/resolver"
+)
+
+var deleteCmd = common.StandardCommand(&cobra.Command{
+	Use:     "delete <deployment ID> [--fail]",
+	Short:   "Delete inactive deployment",
```
Comment (efiShtain) [inline cmd/ak/cmd/deployments/delete.go:12]: technically the delete code above drains a deployment, so it deletes active deployments as well

File: cmd/ak/cmd/deployments/delete.go [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+	RunE: func(cmd *cobra.Command, args []string) error {
+		r := resolver.Resolver{Client: common.Client()}
+		d, id, err := r.DeploymentID(args[0])
+		if err != nil {
+			return common.FailIfError(cmd, err, "deployment")
+		}
+
+		if err := common.FailIfNotFound(cmd, "deployment", d); err != nil { // test d != nil
+			return err
+		}
```
Comment (efiShtain) [inline cmd/ak/cmd/deployments/delete.go]: do we really need to check if deployments exists
and then try to delete it ?
we will get the same result by just calling delete if it exists it will be deleted (assuming it can reach inactive state)
if it is not found, we will get not found error

File: proto/autokitteh/sessions/v1/svc.proto [hunk 15]
```diff
@@ -61,6 +61,12 @@ message GetLogResponse {
   SessionLog log = 1 [(buf.validate.field).required = true];
 }
 
+message DeleteRequest {
+  string session_id = 1 [(buf.validate.field).string.min_len = 1];
+}
+
+message DeleteResponse {}
+
```
Comment (efiShtain) [inline proto/autokitteh/sessions/v1/svc.proto:69]: is this part of this PR ?

File: sdk/sdkclients/sdkdeploymentsclient/client.go [hunk 16]
```diff
@@ -20,25 +20,26 @@ type client struct {
...
 
-// Get implements sdkservices.Deployments.
-func (c *client) Get(ctx context.Context, id sdktypes.DeploymentID) (sdktypes.Deployment, error) {
-	resp, err := c.client.Get(ctx, connect.NewRequest(&deploymentsv1.GetRequest{DeploymentId: id.String()}))
+func New(p sdkclient.Params) sdkservices.Deployments {
+	return &client{client: internal.New(deploymentsv1connect.NewDeploymentsServiceClient, p)}
+}
+
+// Create implements sdkservices.Deployments.
+func (c *client) Create(ctx context.Context, deployment sdktypes.Deployment) (sdktypes.DeploymentID, error) {
```
Comment (efiShtain) [inline sdk/sdkclients/sdkdeploymentsclient/client.go:28]: did we change the API here intentionally to return just the id and not the whole model?

File: sdk/sdkclients/sdkdeploymentsclient/client.go [hunk 17]
```diff
@@ -69,41 +70,44 @@ func (c *client) Test(ctx context.Context, id sdktypes.DeploymentID) error {
...
 
-// Create implements sdkservices.Deployments.
-func (c *client) Create(ctx context.Context, deployment sdktypes.Deployment) (sdktypes.DeploymentID, error) {
-	resp, err := c.client.Create(ctx, connect.NewRequest(&deploymentsv1.CreateRequest{Deployment: deployment.ToProto()}))
+// Drain implements sdkservices.Deployments.
+func (c *client) Drain(ctx context.Context, id sdktypes.DeploymentID) error {
+	resp, err := c.client.Drain(ctx, connect.NewRequest(&deploymentsv1.DrainRequest{DeploymentId: id.String()}))
 	if err != nil {
-		return nil, rpcerrors.TranslateError(err)
+		return rpcerrors.TranslateError(err)
```
Comment (efiShtain) [inline sdk/sdkclients/sdkdeploymentsclient/client.go:77]: can probably add a FromConnectError matching the ToConnectError to have it more zen
WDYT ?

File: backend/internal/db/dbgorm/gorm_test.go (hunk 5 above)
Comment (ig-ra) [reply backend/internal/db/dbgorm/gorm_test.go]: testing. otherwise I need to create deployment, build, env for everything.
And there is specific test that ensures that foreign keys are working

File: backend/internal/db/dbgorm/deployments.go (hunk 4 above)
Comment (ig-ra) [reply backend/internal/db/dbgorm/deployments.go]: I'm waiting for Itays new generic implementation. If this is something that would allow me to solve this, then good. Otherwise I may consider either remove TODO or add a ticket

File: backend/internal/db/dbgorm/builds.go (hunk 3 above)
Comment (ig-ra) [reply backend/internal/db/dbgorm/builds.go]: No, I did it due to my fix for foreign keys. Till now foreign keys weren't enforced for Deployment/Sesions. See Notes in the text. That's why later I've introduce the tests which don't require foreign keys

File: backend/internal/db/dbgorm/deployments.go (hunk 2 above)
Comment (ig-ra) [reply backend/internal/db/dbgorm/deployments.go]: I  left the comment for myself. Once I'm finishing with deleting of everything (connecitons, projects etc) I will decide whether I need to keep generic delete or remove it. Right now we are not using it in deployments, only in session

File: backend/internal/db/dbgorm/deployments.go (hunk 1 above)
Comment (ig-ra) [reply backend/internal/db/dbgorm/deployments.go]: see schema. Deployment cascading soft-delete to sessions

File: cmd/ak/cmd/deployments/delete.go (hunk 13 above)
Comment (ig-ra) [reply cmd/ak/cmd/deployments/delete.go:12]: explain?

File: cmd/ak/cmd/deployments/delete.go (hunk 14 above)
Comment (ig-ra) [reply cmd/ak/cmd/deployments/delete.go]: I'm following the logic of other cmd commands. If you think it's not needed, let's discuss and open a ticket to fix all commands?

File: proto/autokitteh/sessions/v1/svc.proto (hunk 15 above)
Comment (ig-ra) [reply proto/autokitteh/sessions/v1/svc.proto:69]: explain?

File: sdk/sdkclients/sdkdeploymentsclient/client.go (hunk 17 above)
Comment (ig-ra) [reply sdk/sdkclients/sdkdeploymentsclient/client.go:77]: sure, worth a ticker. This is the first step into this direction indeed. But we are still not there. Need to fix tests, then write tests for all commands and then ensure the right errors. 
I think I will finish with all commands first

File: backend/internal/deployments/deployments.go (hunk 12 above)
Comment (ig-ra) [reply backend/internal/deployments/deployments.go]: this is the solution. My claim that we could delete inactive deployment not in transaction

File: backend/internal/db/dbgorm/gorm_test.go (hunk 6 above)
Comment (ig-ra) [reply backend/internal/db/dbgorm/gorm_test.go]: yes, I will move all tests to use formkitteh. But not now. I will open a ticket
https://linear.app/autokitteh/issue/ENG-476/move-db-tests-to-use-gormkitteh

File: backend/internal/db/dbgorm/gorm_test.go (hunk 7 above)
Comment (ig-ra) [reply backend/internal/db/dbgorm/gorm_test.go]: I will need this in next PRs. That's why it's still there

File: backend/internal/db/dbgorm/scheme/scheme.go (hunk 8 above)
Comment (ig-ra) [reply backend/internal/db/dbgorm/scheme/scheme.go]: I think it's written in comment above. Without having Sessions reference in Deployment schema, gorm won't enforce it and will allow to add sessions w/o matching deployment. This was a bug I solved.

File: backend/internal/db/dbgorm/scheme/scheme.go (hunk 9 above)
Comment (ig-ra) [reply backend/internal/db/dbgorm/scheme/scheme.go]: I prefer to be explicit in this case. But I think I could remove references.
What do u prefer?

File: backend/internal/db/dbgorm/scheme/scheme.go (hunk 10 above)
Comment (ig-ra) [reply backend/internal/db/dbgorm/scheme/scheme.go]: This is DB implementation. The logic of checking whether we could delete deployment isn't here but in deployments.Delete. There I'm allowing to delete only inactive deployment. (drain->deactivate->delete). Which ensures that there are no sessions.

File: backend/internal/db/dbgorm/sessions.go (hunk 11 above)
Comment (ig-ra) [reply backend/internal/db/dbgorm/sessions.go]: nil

File: sdk/sdkclients/sdkdeploymentsclient/client.go (hunk 16 above)
Comment (ig-ra) [reply sdk/sdkclients/sdkdeploymentsclient/client.go:28]: I don't think I've changed any API here. My diff shows me that Create was returning ID. Am I wrong?

File: backend/internal/db/dbgorm/scheme/scheme.go (hunk 8 above)
Comment (efiShtain) [reply backend/internal/db/dbgorm/scheme/scheme.go]: the relation we have is session belongs to deployment
we didn't need deployment has sessions
do we need the two way? what changes in the database after this change ?

File: backend/internal/db/dbgorm/scheme/scheme.go (hunk 9 above)
Comment (efiShtain) [reply backend/internal/db/dbgorm/scheme/scheme.go]: minimum of options are better
I think it can even work without any of them

File: cmd/ak/cmd/deployments/delete.go (hunk 13 above)
Comment (efiShtain) [reply cmd/ak/cmd/deployments/delete.go:12]: you say delete inactive deployments in the text
but delete allows deleting active deployment, it just moves it to inactive first by draining it

File: sdk/sdkclients/sdkdeploymentsclient/client.go (hunk 16 above)
Comment (efiShtain) [reply sdk/sdkclients/sdkdeploymentsclient/client.go:28]: you are right, 
I read it wrong

File: proto/autokitteh/sessions/v1/svc.proto (hunk 15 above)
Comment (efiShtain) [reply proto/autokitteh/sessions/v1/svc.proto:69]: why there are changes to the sessions svc ?
you added delete to the deployments svc

File: cmd/ak/cmd/deployments/delete.go (hunk 13 above)
Comment (ig-ra) [reply cmd/ak/cmd/deployments/delete.go:12]: you are right. I already asked this question Haim and Itay. Got no answer meanwhile

Comment (ig-ra) [reply cmd/ak/cmd/deployments/delete.go:12]: rewritten. Now we are checking that deployment is inactive and then deleting, err otherwise

File: backend/internal/db/dbgorm/scheme/scheme.go (hunk 8 above)
Comment (ig-ra) [reply backend/internal/db/dbgorm/scheme/scheme.go]: DB wont be changed. But without having sessions in Deployment foreign keys won't be forced by gorm.
Gorm is enforcing foreign keys only if you have relevant objects/pointers in the schema. 
It's semi-stated here https://gorm.io/docs/belongs_to.html, https://gorm.io/docs/has_many.html
see also https://medium.com/@chaewonkong/learn-go-understanding-and-implementing-foreign-keys-with-gorm-6d7608e1dbf6

File: proto/autokitteh/sessions/v1/svc.proto (hunk 15 above)
Comment (ig-ra) [reply proto/autokitteh/sessions/v1/svc.proto:69]: just moved to the upper part of the file. refactoring

File: backend/internal/db/dbgorm/scheme/scheme.go (hunk 9 above)
Comment (ig-ra) [reply backend/internal/db/dbgorm/scheme/scheme.go]: deleted references

**Review by itayd — APPROVED** [review]
preemptive to ungate. either efi or daniel must approve as well.
approved by Efi

**Review by efiShtain — APPROVED** [review]

File: cmd/ak/cmd/deployments/deployments.go [hunk 18]
```diff
@@ -30,6 +30,7 @@ func init() {
 	deploymentsCmd.AddCommand(deactivateCmd)
 	deploymentsCmd.AddCommand(getCmd)
 	deploymentsCmd.AddCommand(listCmd)
+	deploymentsCmd.AddCommand(deleteCmd)
```
Comment (daabr) [inline cmd/ak/cmd/deployments/deployments.go]: Cobra displays commands alphabetically, let's do the same when we add subcommands to a parent (I know we don't always do it yet, but we will).

File: cmd/ak/cmd/deployments/delete.go [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/spf13/cobra"
+
+	"go.autokitteh.dev/autokitteh/cmd/ak/common"
+	"go.autokitteh.dev/autokitteh/internal/resolver"
+)
+
+var deleteCmd = common.StandardCommand(&cobra.Command{
+	Use:     "delete <deployment ID> [--fail]",
+	Short:   "Delete inactive deployment",
+	Aliases: []string{"d"},
```
Comment (daabr) [inline cmd/ak/cmd/deployments/delete.go]: We already have 2 other commands that start with D: "drain" (dr), and "deactivate" (de). Please remove the alias "d", or change it to "del"?

File: cmd/ak/cmd/deployments/delete.go [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+			return common.FailIfError(cmd, err, "deployment")
+		}
+
+		if err := common.FailIfNotFound(cmd, "deployment", d); err != nil { // test d != nil
+			return err
+		}
+
+		ctx, cancel := common.LimitedContext()
+		defer cancel()
+		err = deployments().Delete(ctx, id)
```
Comment (daabr) [inline cmd/ak/cmd/deployments/delete.go]: I'm not a big fan of empty lines like Itay, but I do see a need to emphasize that things with `defer` are "blocks" of code, no matter how small.

```suggestion
		defer cancel()

		err = deployments().Delete(ctx, id)
```

File: cmd/ak/cmd/deployments/delete.go [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+		if err != nil { // don't ignore silently any error, expect not found above
+			return common.ToExitCodeError(err, "delete deployment")
```
Comment (daabr) [inline cmd/ak/cmd/deployments/delete.go]: `common.ToExitCodeError` already handles `err == nil`, why do you need this?

**Review by daabr — APPROVED** [review]

File: cmd/ak/cmd/deployments/delete.go (hunk 20 above)
Comment (ig-ra) [reply cmd/ak/cmd/deployments/delete.go]: done

File: cmd/ak/cmd/deployments/deployments.go (hunk 18 above)
Comment (ig-ra) [reply cmd/ak/cmd/deployments/deployments.go]: done

File: cmd/ak/cmd/deployments/delete.go (hunk 19 above)
Comment (ig-ra) [reply cmd/ak/cmd/deployments/delete.go]: done
